Multipart pipe reader

gilmishal opened the pull request:
Continuing work on MultipartForm PipeReader Implementation from PR #15056.

Addresses #11377

Commit:
added multipart pipe reader and tests from Tratcher's branch

Commit:
fixed failing unit tests

gilmishal commented on src/Http/WebUtilities/test/MultipartPipeReaderTests.cs:
```diff
@@ -0,0 +1,376 @@
+        }
+
+        [Fact]
+        public async Task MultipartPipeReader_TwoPartBodyIncompleteBuffer_OneSectionReadsSuccessfullySecondSectionThrows()
```
I fixed the test to pass, but I am not sure if it's a functional requirement or not - Adding a subsequent AdvanceTo and ReadAsync will throw, but I am not sure if that's what we want. 
Basically, it's conflicting with a comment inside the MultipartSectionPipeReader  _We want them to consume and advance all of the data before we try to drain the metadata / boundary._

gilmishal replied:
Looking at the original now, I am not sure I would say that the behavior had changed.
If we alter the original test and add another read to the section.Body it would fail.
If we alter this test to read the third section instead of continuing to read the rest of the second section it would also fail.

I think that this behavior was changed with my original implementation of the PipeReader, and the test just wasn't updated.

Anyways, we might want to consider altering the test to reflect the actual scenario - basically the behavior in both the Stream implementation and the Pipe Implementation is that for an incomplete Section Boundry - the second section will throw before reaching the end - (Pipe - IsCompleted == true, Stream - Read returns 0) - and trying to read the third section will throw.  I would argue that in reality we are more likely to try and read the second section to it's end before attempting to read the third section, especially since we will probably copy each section to it's own Pipe.

gilmishal commented on src/Http/WebUtilities/src/MultipartSectionPipeReader.cs:
```diff
@@ -73,11 +73,21 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
+                {
+                    // We reached the end of the section, we should advance to the end.
+                    _pipeReader.AdvanceTo(_unconsumedData.End);
+                    _finalLineConsumed = true;
```
This solves an issue with _finalLineConsumed is false (and probably shouldn't be), I am not sure if it's the correct solution though, and I might be missing an edge case.

gilmishal commented on src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs:
```diff
@@ -0,0 +1,197 @@
+        [InlineData(TextAndBoundary, Text, 1)]
+        [InlineData(HtmlWithNewLinesAndBoundary, HtmlWithNewLines, 52)]
+        [InlineData(TextWithPartialBoundaryMatchAndBoundary, TextWithPartialBoundaryMatch, 14)]
+        public async Task MultipartSectionPipeReader_SingleByteReadWithoutAdvance_Success(string input, string expected, int reads)
```
Not sure how it differs from  WithAdvance Test. The only difference is that the result isn't buffered and the expected bytes read is larger - both seem like a mistake in the test rather than the functionality, since I am not sure how it's even supposed to pass - **I am gonna guess that this is a copy pasted test that was never implemented, but I am double checking.**

Tratcher replied:
The difference is that the prior test always calls `AdvanceTo(temp.Buffer.End);` but this one calls `AdvanceTo(temp.Buffer.Start, temp.Buffer.End);`. That makes the former unbuffered and the latter buffered. Both are valid scenarios.

gilmishal replied:
Oh, I might have missed that. Anyways, I was wondering why they were expecting different number of reads, since they are both pretty similar. I am guessing that this was a mistake.

gilmishal commented on src/Http/WebUtilities/test/MultipartSectionPipeReaderTests.cs:
```diff
@@ -0,0 +1,197 @@
+            var sectionReader = new MultipartSectionPipeReader(pipeReader, new MultipartBoundary(Boundary));
+
+            var result = await sectionReader.ReadAsync();
+            Assert.False(result.IsCompleted);
```
IsCompleted currently comes out as true, This is because it buffered the entire section on the first Read to the underlying Pipe, and it truly reached the end of the section. I am not sure if that's a problem or not.

Tratcher replied:
It's fine for IsComplete to return true so long as all of the data is available and we've fully parsed the boundary. Just make sure AdvanceTo(end) also handles advancing the underlying pipe past the parsed boundary.

gilmishal replied:
Ok then in that case, I think the test is fixed and covers this case.

gilmishal commented:
@Tratcher I have made most  of the unit tests pass - there are some tests that are still failing, but I need your help deciding what to do with these tests. 

I feel that the Pipe Implementation itself is solid (not sure about the test coverage though), but we still need to decide how to buffer (and deliver) the sections, to allow reading the sections interchangeably.

En3Tho commented:
Can you post benchmark results here please? Just to see the effect in plain numbers

Tratcher reviewed (commented):
I've only skimmed the code and answered questions so far.

gilmishal commented:
> Can you post benchmark results here please? Just to see the effect in plain numbers

I am unable to run any of the benchmarking projects - I am getting this error 
`Project Microsoft.AspNetCore.WebUtilities.Performance is not compatible with net5.0`

I am not sure what's causing it, and I couldn't find any relevant documentation in this repo, or on BenchmarkDotNet.

I can say that I did run activate.ps1 assuming it was related to the cli dotnet version being used (apperantly not though) - but I do think it might be related to this lines in Program.cs (in the performance projects)

 ```
.WithToolchain(CsProjCoreToolchain.From(new NetCoreAppSettings
                    (
                        // not using "net6.0", a workaround for https://github.com/dotnet/BenchmarkDotNet/pull/1479
                        targetFrameworkMoniker: "netcoreapp5.0",
                        runtimeFrameworkVersion: default,
                        name: ".NET Core 5.0"
                    )))
```

I don't have any other angle at the moment, other than perhaps trying to build this thing for .net 5.

gilmishal committed:
made all tests pass

gilmishal commented on src/Http/WebUtilities/src/MultipartPipeSection.cs:
```diff
@@ -0,0 +1,59 @@
+        /// The position where the body starts in the total multipart body.
+        /// This may not be available if the total multipart body is not seekable.
+        /// </summary>
+        public long? BaseStreamOffset { get; set; }
```
removed

gilmishal commented:
I have addressed your comments and fixed the tests accordingly, please verify that you approve my changes, and with your permission I will fix TryRead and create the tests for it.
Then, I will start working on buffering the sections.

gilmishal committed:
Added XML Comments, removed unused code

gilmishal commented on src/Http/WebUtilities/src/MultipartSectionPipeReader.cs:
```diff
@@ -180,7 +180,7 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
                 {
                     // We must be examining a partial boundary match, keep reading.
-                    _pipeReader.AdvanceTo(_unconsumedData.Start, _unconsumedData.End);
+                    _pipeReader.AdvanceTo(readResult.Buffer.Start, _unconsumedData.End);
```
This was originally _bodyBuffer.Start - I changed it to _unconsumedData.Start, which is totally wrong because it advanced the pipe without the user advancing it.

The reason I changed it was that _bodyBuffer.Start was resulting in a never ending loop - that happend because if the read isn't long enough to check if it's a boundary - it wouldn't set the _bodyBuffer (which does make sense)  - since _bodyBuffer was unset, the advanceTo wouldn't actually advance, and create a never ending loop.

I think that readResult.Buffer.Start would avoid this issue, and reflects the actual meaning (that's why I changed all advances to use it instead of _bodyBuffer.Start - Just informing you so you find it easier to understand the changes I made.

davidfowl commented:
👀

Tratcher commented on src/Http/WebUtilities/src/MultipartPipeSection.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+
```
```suggestion
```

Tratcher commented on src/Http/WebUtilities/perf/Microsoft.AspNetCore.WebUtilities.Performance/MultipartReaderBenchmark.cs:
```diff
@@ -0,0 +1,61 @@
+{
+    public class MultipartReaderBenchmark
+    {
+
```
```suggestion
```

Tratcher reviewed (commented):
Here's a little more detailed feedback. I didn't get to the most complex part in MultipartSectionPipeReader.

I talked this over with our team and we reached two conclusions:
1) We're starting our 6.0 development and focusing on P0 features for the first milestone or two. We do think this multipart pipe reader is important for 6.0, but it will be a milestone or two before we're able to focus on it.
2) @jkotalik has a lot more experience with custom pipe readers so he'll take over reviewing this PR.

gilmishal committed:
Apply suggestions from code review

addressed todo regarding HeadersLengthLimit

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

gilmishal committed:
few cosmetic fixes

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

gilmishal commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,331 @@
+                try
+                {
+                    var finishedParsing = TryParseHeadersToEnd(ref buffer, ref headersAccumulator, ref headersLength);
+                    if (headersLength > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                    }
```
yes, it's redundant I'll remove.

gilmishal commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,331 @@
+                }
+                finally
+                {
+                    _pipeReader.AdvanceTo(buffer.Start, buffer.End);
```
the buffer is being sliced within TryParseHeadersToEnd

jkotalik replied:
I still think there could be issues here. I think the code path for TryParseHeaderstoEndFast is fine, but I have concerns about TryParseheadersToEndSlow. At the end, you are doing `buffer = default`, which I think needs to be `buffer = buffer.Slice(sequenceReader.Position);`

gilmishal replied:
I think you are right. This certainly seems like a possible issue, I will also add a test for coverage.

gilmishal replied:
Definitely an issue, I managed to reproduce it when removing the ParseHeadersFast method and using a SingleByteStream - there were duplicate header values - updated the test as much as I could.
I do suggest adding tests forcing ParseHeadersSlow - but I am not sure how to do it, I don't know how to force multi segment - and I wouldn't want to change functionalities just for the test. Anyways the issue itself is fixed by your suggestion.

gilmishal committed:
removed redundant length check, fixed nullable type

jkotalik commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+                }
+
+                var lineReader = new SequenceReader<byte>(line);
+                ReadOnlySequence<byte> value;
```
No need to declare value early.

gilmishal replied:
done

jkotalik commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+                    throw new InvalidDataException($"Invalid header line: {GetStringFromReadOnlySequence(line)}");
+                }
+                value = line.Slice(lineReader.Position);
+
```
```suggestion
```

gilmishal replied:
done

jkotalik commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+                    // Don't buffer indefinitely
+                    if (headersLength + sequenceReader.Consumed > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {HeadersLengthLimit - headersLength} exceeded.");
```
```suggestion
                    throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
```

jkotalik commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+            }
+
+            buffer = default;
```
Does this code path only execute when the sequence is empty?

gilmishal replied:
It executes when the buffer is empty, or if we read till the end of the buffer, but didn't find all the headers - and didn't exceed the Headers' length limit.

jkotalik commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+            if (_currentSectionReader == null) // first call.
+            {
+                // This section will drain any preamble data and remove the first boundary marker.
+                _currentSectionReader = new MultipartSectionPipeReader(_pipeReader, _boundary) { LengthLimit = HeadersLengthLimit };
```
Idea (for future), could we get away with not having the MultipartSectionPipeReader class at all, tracking state in the MultipartPipeReader?

jkotalik reviewed (commented):
Some initial comments. Still trying to get my head around the MultipartSectionPipeReader.

gilmishal commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+
+            if (_currentSectionReader.FinalBoundaryFound)
+            {
+                return null;
```
In the pipeReader implementation we are reading to the end of the pipeReader once we find the finalBoundary, so in practice the behavior is the same, and there is no need to drain the trailing data - basically the difference is that the pipe is just drained to end as soon as the final boundary is found , instead of waiting for a subsequent DrainAsync.

https://github.com/dotnet/aspnetcore/blob/b784d8ea2d34513dd12a1f533a17431679c22c63/src/Http/WebUtilities/src/MultipartSectionPipeReader.cs#L187-L195

gilmishal commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+
+                if (line.Length > lengthLimit)
+                {
+                    throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
```
I suspect these exceptions are the way they are since I tried sticking to the original behavior, which seems like a limitation from the way it was implemented with the Buffered Read Stream -

I don't think there is in fact a length limit for a specific header, and think it would make sense to use the same error message in all these situations and just use `throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");`

gilmishal replied:
And obviously update the tests I was reluctant to update

gilmishal committed:
Addressed comments, added coverage for a scenario that we missed

gilmishal commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+                ReadOnlySpan<byte> line;
+                var foundNewLine = newLine != -1;
+
+                if (foundNewLine)
```
done

gilmishal commented on src/Http/WebUtilities/src/MultipartPipeReader.cs:
```diff
@@ -0,0 +1,329 @@
+
+                    if (readResult.IsCompleted)
+                    {
+                        throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
```
done

gilmishal commented:
@jkotalik any more comments? Can I continue to the next steps?

jkotalik commented:
@gilmishal apologies, I'll get back to this within the week.

jkotalik commented on src/Http/WebUtilities/src/MultipartPipeSection.cs:
```diff
@@ -0,0 +1,64 @@
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// a Pipe Implementation of a Http Multipart Section
```
```suggestion
    /// A Pipe Implementation of a Http Multipart Section
```

jkotalik commented on src/Http/WebUtilities/src/MultipartSectionPipeReader.cs:
```diff
@@ -0,0 +1,443 @@
+{
+    internal class MultipartSectionPipeReader : PipeReader
+    {
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
```
See https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs for reference to another custom PipeReader implemented (doesn't directly implement PipeReader, but has the same methods).

jkotalik reviewed (commented):
It's hard to follow what's going on in ReadAsync/TryRead/AdvanceTo. One thing I found is writing an example in code with a boundary and end-of-section to show how each of the cases are covered helps with understanding.

jkotalik commented:
@gilmishal you get the chance to respond to this feedback?

gilmishal commented:
Sorry, I'll respond tomorrow, been a crazy month.

gilmishal commented:
> @gilmishal you get the chance to respond to this feedback?

Been a crazy month. I'll respond tomorrow.

jkotalik commented:
I'm going to close this for now, feel free to reopen when you have the chance to review and respond to feedback.
